### PR TITLE
interpret: make overflowing binops just normal binops

### PR DESCRIPTION
**Pull request by RalfJung**
Follow-up to https://github.com/rust-lang/rust/pull/125173 (Cc @scottmcm)

### CODE REVIEW

File: compiler/rustc_mir_transform/src/gvn.rs
Comment (RalfJung) [inline compiler/rustc_mir_transform/src/gvn.rs:226]: I think there's more clean-up possible here (and in dataflow_const_prop as well), but I'll leave that to someone else.

Cc @cjgillot

File: compiler/rustc_const_eval/src/interpret/operator.rs
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/operator.rs:44]: Removing this code means we now assume that the return type of the WithOverflow ops actually is a ScalarPair. I think we should enforce that our primitive operations return data we can represent with out primitive types, so IMO this is a reasonable assumption. At some point we added the codepath here to account for layout randomization, but looking through the layout code I think tuples are using `ReprFlags::default()` which will not have `RANDOMIZE_LAYOUT` set so this shouldn't happen any more?

If we ever make it happen again, we should not randomize tuples of the form `($int, bool)`.

File: compiler/rustc_mir_transform/src/dataflow_const_prop.rs
Comment (RalfJung) [inline compiler/rustc_mir_transform/src/dataflow_const_prop.rs:269]: I don't understand *why* the overflowing operators can't be handled here, but it matches the previous structure where `Rvalue::CheckedBinaryOp` was handled in `handle_assign`, not in `handle_rvalue`, for some reason.

**Review by scottmcm — APPROVED** [review]
I did a (admittedly fairly quick) scan over this and it seems good to me, but you should probably get a review from someone more familiar with the nuances of CTFE/Miri.

File: compiler/rustc_const_eval/src/interpret/operator.rs
Comment (RalfJung) [reply compiler/rustc_const_eval/src/interpret/operator.rs:44]: That's for `Rvalue::Aggregate`, I don't see how it is related.

But yeah we could make `ScalarPair` more flexible, we just need to add the information at which offset the two fields are living. Then all variants of `($int, bool)` can be represented. (Also see [here](https://github.com/rust-lang/rust/issues/119183) and [here](https://github.com/rust-lang/compiler-team/issues/672#issuecomment-1751624436) for the more general topic of refactoring our `Abi` handling.)

File: compiler/rustc_middle/src/mir/tcx.rs
Comment (RalfJung) [reply compiler/rustc_middle/src/mir/tcx.rs:298]: Yeah I have no idea.^^ It doesn't have a doc comment either.

But we lower the HIR `+` to `AddWithOverflow` in debug builds so IMO it makes sense to add that backwards mapping here. :shrug:

**Comment by RalfJung** [comment]
> you should probably get a review from someone more familiar with the nuances of CTFE/Miri.

r? @oli-obk

**Comment by oli-obk** [comment]
@bors try @rust-timer queue

r=me with perf clean. I'll give the known panics pass another look for more cleanups once this lands

**Comment by RalfJung** [comment]
FWIW I think GVN/const-prop have the more obvious cleanup possibilities. But yeah known-lints can probably also be cleaned up a bit more.

**Comment by RalfJung** [comment]
Seems clean enough to me -- a barely significant regression that only affects some build configurations of the stress test.

File: compiler/rustc_mir_transform/src/dataflow_const_prop.rs
Comment (cjgillot) [reply compiler/rustc_mir_transform/src/dataflow_const_prop.rs:269]: As checked ops return scalar pairs, the routine needs to project the destination as 2 scalars, which `handle_rvalue` is not designed to do.

Comment (RalfJung) [reply compiler/rustc_mir_transform/src/dataflow_const_prop.rs:269]: Would be good to add a comment in the code explaining that.

**Pull request merged**
